File: geo/collect.py
"""Stage 1 — collection. Fan the prompt set across engines, store raw answers."""
from __future__ import annotations

import json
import os
import time
import uuid

from tqdm import tqdm

from .config import AppConfig
from .engines import get_engine
from .prompts import read_prompts_csv


def collect(cfg: AppConfig, limit: int | None = None) -> list[dict]:
    prompts = read_prompts_csv(cfg.prompts_path)
    if limit:
        prompts = prompts[:limit]

    # resolve engines once; warn + skip unavailable ones
    engines = []
    for name in cfg.engines:
        eng = get_engine(name, cfg)
        if eng is None:
            print(f"[collect] engine '{name}' has no adapter — skipping")
        elif not eng.available:
            print(f"[collect] engine '{name}' unavailable (missing key) — skipping")
        else:
            engines.append(eng)
    if not engines:
        raise RuntimeError("No available engines. Set ANTHROPIC_API_KEY and ENGINES.")

    records: list[dict] = []
    total = len(prompts) * len(engines) * cfg.responses_per_prompt
    bar = tqdm(total=total, desc="Collecting answers", unit="ans")

    for row in prompts:
        prompt = row["prompt"]
        for eng in engines:
            for run_i in range(cfg.responses_per_prompt):
                res = eng.answer(prompt)
                records.append({
                    "uuid": str(uuid.uuid4()),
                    "engine": res.engine,
                    "model": res.model,
                    "run": run_i,
                    "prompt": prompt,
                    "persona": row.get("persona", ""),
                    "topic": row.get("topic", ""),
                    "intent": row.get("intent", ""),
                    "response": res.text,
                    "cited_urls": res.cited_urls,
                    "searched_urls": res.searched_urls,
                    "error": res.error,
                    "meta": res.meta,
                })
                bar.update(1)
                if res.error:
                    bar.write(f"[collect] {res.engine} error on '{prompt[:50]}...': {res.error}")
                if cfg.rate_limit_sleep:
                    time.sleep(cfg.rate_limit_sleep)
    bar.close()

    os.makedirs(cfg.output_dir, exist_ok=True)
    out = os.path.join(cfg.output_dir, "responses.json")
    with open(out, "w", encoding="utf-8") as f:
        json.dump(records, f, ensure_ascii=False, indent=2)
    print(f"[collect] wrote {len(records)} answers -> {out}")
    return records

File: web/lib/pipeline/collect.ts
// Stage 1 — collection. Fan the prompt set across engines, store raw answers.
// Ported from geo/collect.py. Adds bounded concurrency for speed.
import path from "node:path";
import { AppConfig } from "./config";
import { getEngine, Engine } from "./engines";
import { readPromptsCsv } from "./prompts";
import { writeJson } from "./io";
import type { Row } from "./metrics";

export type Progress = (msg: string) => void;

function randomUuid(): string {
  // Node 18+ has global crypto.randomUUID
  try {
    return (globalThis.crypto as any).randomUUID();
  } catch {
    return "id-" + Math.random().toString(36).slice(2);
  }
}

async function pool<T, R>(items: T[], limit: number, fn: (item: T, i: number) => Promise<R>): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let next = 0;
  const workers = Array.from({ length: Math.min(limit, items.length) }, async () => {
    while (true) {
      const i = next++;
      if (i >= items.length) break;
      results[i] = await fn(items[i], i);
    }
  });
  await Promise.all(workers);
  return results;
}

export async function collect(
  cfg: AppConfig,
  limit: number | null,
  onProgress?: Progress,
  concurrency = 4,
): Promise<Row[]> {
  let prompts = readPromptsCsv(cfg.promptsPath);
  if (limit) prompts = prompts.slice(0, limit);

  const engines: Engine[] = [];
  for (const name of cfg.engines) {
    const eng = getEngine(name, cfg);
    if (eng == null) onProgress?.(`[collect] engine '${name}' has no adapter — skipping`);
    else if (!eng.available()) onProgress?.(`[collect] engine '${name}' unavailable (missing key) — skipping`);
    else engines.push(eng);
  }
  if (!engines.length) throw new Error("No available engines. Set ANTHROPIC_API_KEY and ENGINES.");

  // Flatten to (prompt, engine, run) units so we can pool across them.
  type Unit = { row: (typeof prompts)[number]; eng: Engine; run: number };
  const units: Unit[] = [];
  for (const row of prompts) {
    for (const eng of engines) {
      for (let run = 0; run < cfg.responsesPerPrompt; run++) units.push({ row, eng, run });
    }
  }

  const total = units.length;
  let done = 0;
  // Format includes a `| done/total [` segment so the loading screen's tqdm-style
  // parser (web/app/setup/loading) can extract progress.
  onProgress?.(`Collecting answers: 0%| 0/${total} [ans]`);

  const records = await pool(units, concurrency, async (u) => {
    const res = await u.eng.answer(u.row.prompt);
    done++;
    if (done % Math.max(1, Math.floor(total / 50)) === 0 || done === total) {
      const pct = Math.floor((100 * done) / total);
      onProgress?.(`Collecting answers: ${pct}%| ${done}/${total} [ans]`);
    }
    if (res.error) onProgress?.(`[collect] ${res.engine} error on '${u.row.prompt.slice(0, 50)}...': ${res.error}`);
    return {
      uuid: randomUuid(),
      engine: res.engine,
      model: res.model,
      run: u.run,
      prompt: u.row.prompt,
      persona: u.row.persona || "",
      topic: u.row.topic || "",
      intent: u.row.intent || "",
      response: res.text,
      cited_urls: res.cited_urls,
      searched_urls: res.searched_urls,
      error: res.error,
      meta: res.meta,
    } as Row & { uuid: string; run: number; meta: unknown };
  });

  const out = path.join(cfg.outputDir, "responses.json");
  writeJson(out, records);
  onProgress?.(`[collect] wrote ${records.length} answers -> ${out}`);
  return records as unknown as Row[];
}
